watidl/wasi-core.watidl

```diff
@@ -0,0 +1,1076 @@
+		;; List of scatter/gather vectors in which to store data.
+		(param $iovs $wasi.iovec_t_array)
+		;; The offset within the file at which to read.
+		(param $offset $wasi.filesize_t)
+		;; The number of bytes read.
+		(result $nread $wasi.size_t)
+	)
+
+	;; Return a description of the given preopened file descriptor.
+	(func (export "prestat_get")
+		;; The buffer where the description is stored.
+		(result $buf $wasi.prestat_t)
+	)
+
+	;; Return a description of the given preopened file descriptor.
+	(func (export "prestat_dir_name")
+		;; A buffer into which to write the preopened directory name.
+		(param $path string)
+	)
+
+	;; Write to a file descriptor, without using and updating the file descriptor's offset.
+	;; Note: This is similar to `pwritev` in POSIX.
+	(func (export "pwrite")
+		;; List of scatter/gather vectors from which to retrieve data.
+		(param $iovs $wasi.iovec_t_array)
+		;; The offset within the file at which to write.
+		(param $offset $wasi.filesize_t)
+		;; The number of bytes written.
+		(result $nwritten $size_t)
+	)
+
+	;; Read from a file descriptor.
+	;; Note: This is similar to `readv` in POSIX.
+	(func (export "read")
+		;; List of scatter/gather vectors to which to store data.
+		(param $iovs $wasi.iovec_t_array)
+		;; The number of bytes read.
+		(result $nread $size_t)
+	)
+
+	;; Read directory entries from a directory.
+	;; When successful, the contents of the output buffer consist of a sequence of directory entries. Each directory entry consists of a dirent_t object, followed by dirent_t::d_namlen bytes holding the name of the directory entry.
+	;; This function fills the output buffer as much as possible, potentially truncating the last directory entry. This allows the caller to grow its read buffer size in case it's too small to fit a single large directory entry, or skip the oversized directory entry.
+	(func (export "readdir")
+		;; The buffer where directory entries are stored
+		(param $buf data)
+		;; The location within the directory to start reading
+		(param $cookie $wasi.dircookie_t)
+		;; The number of bytes stored in the read buffer. If less than the size of the read buffer, the end of the directory has been reached.
+		(result $bufused $size_t)
+	)
+
+	;; Atomically replace a file descriptor by renumbering another file descriptor.
+	;; Due to the strong focus on thread safety, this environment does not provide a mechanism to duplicate or renumber a file descriptor to an arbitrary number, like dup2(). This would be prone to race conditions, as an actual file descriptor with the same number could be allocated by a different thread at the same time.
+ 	;; This function provides a way to atomically renumber file descriptors, which would disappear if dup2() were to be removed entirely.
+	(func (export "renumber")
+		;; The file descriptor to overwrite.
+		(param $to $fd)
+	)
+
+	;; Move the offset of a file descriptor.
+	;; Note: This is similar to lseek in POSIX.
+	(func (export "seek")
```

Comment:
So is "fd" a separate WASI interface or just part of wasi-core?    Right now we just have wasi-unstable as a single interface.     If we wanted to literally translate the current wasi-unstable as it stands would be just one interface?

Maybe this isn't a precise representation of wasi-unstable as it is today, but if one wanted to do that presumably one wouldn't try to mode file functions as methods on the 'fd' object like this?  This seems like add OO-style when there was none previously.

Reply:
> So is "fd" a separate WASI interface or just part of wasi-core?

Currently its part of WASI core, but I think there some plans to break up wasi-core into several modules. So this is trying to reflect that future. I was thinking of adding `(interface $fd (export "wasi-unstable")` to explicitly show that all of the imports are under the "wasi-unstable" module namespace.  In the future it would be nice to have multiple modules we could use so I imagine common types would be imported from "wasi-core" or just "wasi" base file descriptor methods would be imported from "wasi-fd" ect. 

> Maybe this isn't a precise representation of wasi-unstable as it is today

I think it would be nice if this could represent wasi-unstable today. I think we can do that by adding some notation to that says "all these interfaces have the same module namespace". That way we can introduce this notation without having to make other changes.

>but if one wanted to do that presumably one wouldn't try to mode file functions as methods on the 'fd' object like this?

So there is another idea here. Yes its very OO-style centric and ideally we can having binding that allow wasm programs to have two different modes of interacting with the interface 1) is the current ADT pattern or 2) have some sort of dynamic dispatching which would allow for virtualization. 

The OO-style would suggest number # 2 but can also express # 1 But if don't have some OO-style notation then how would represent # 2 in the future?